lib/ruby_lsp/requests/hover.rb

```diff
@@ -28,119 +30,61 @@ def provider
+            !Listeners::Hover::ALLOWED_TARGETS.include?(target.class)) ||
+            (parent.is_a?(Prism::ConstantPathNode) && target.is_a?(Prism::ConstantReadNode))
+          target = parent
         end
 
-        self
-      end
-
-      sig { params(node: Prism::ConstantReadNode).void }
-      def on_constant_read_node_enter(node)
-        return if DependencyDetector.instance.typechecker
+        @listeners = T.let([], T::Array[Listener[ResponseType]])
 
-        generate_hover(node.slice, node.location)
-      end
-
-      sig { params(node: Prism::ConstantWriteNode).void }
-      def on_constant_write_node_enter(node)
-        return if DependencyDetector.instance.typechecker
+        # Don't need to instantiate any listeners if there's no target
+        return unless target
 
-        generate_hover(node.name.to_s, node.name_loc)
-      end
-
-      sig { params(node: Prism::ConstantPathNode).void }
-      def on_constant_path_node_enter(node)
-        return if DependencyDetector.instance.typechecker
+        @listeners = T.let(
+          [Listeners::Hover.new(nesting, index, dispatcher, typechecker_enabled)],
+          T::Array[Listener[ResponseType]],
+        )
+        Addon.addons.each do |addon|
+          addon_listener = addon.create_hover_listener(nesting, index, dispatcher)
+          @listeners << addon_listener if addon_listener
+        end
 
-        generate_hover(node.slice, node.location)
+        @target = T.let(target, Prism::Node)
+        @dispatcher = dispatcher
       end
 
-      sig { params(node: Prism::CallNode).void }
-      def on_call_node_enter(node)
-        return if @typechecker_enabled
-        return unless self_receiver?(node)
-
-        message = node.message
-        return unless message
-
-        target_method = @index.resolve_method(message, @nesting.join("::"))
-        return unless target_method
-
-        location = target_method.location
-
-        @_response = Interface::Hover.new(
-          range: range_from_location(location),
-          contents: markdown_from_index_entries(message, target_method),
-        )
-      end
+      sig { override.returns(ResponseType) }
+      def response
+        @dispatcher.dispatch_once(@target)
+        responses = @listeners.map(&:response).compact
 
-      private
+        first_response, *other_responses = responses
 
-      sig { params(name: String, location: Prism::Location).void }
-      def generate_hover(name, location)
-        entries = @index.resolve(name, @nesting)
-        return unless entries
+        return unless first_response
 
-        # We should only show hover for private constants if the constant is defined in the same namespace as the
-        # reference
-        first_entry = T.must(entries.first)
-        return if first_entry.visibility == :private && first_entry.name != "#{@nesting.join("::")}::#{name}"
+        # TODO: other_responses should never be nil. Check Sorbet
```

Comment:
Should this comment be removed? I think the `T.must` below addresses it, right?

Reply:
I think the comment means that we shouldn't need to wrap `T.must` around `other_responses` as it's a splat assignment:

```rb
        first_response, *other_responses = responses
```

It should always be an array regardless of `responses`'s value:

```
ruby-lsp(main):008> x, *y = a; y
=> []
ruby-lsp(main):009> a = nil
=> nil
ruby-lsp(main):010> x, *y = a; y
=> []
ruby-lsp(main):011> 
```

Reply:
Yes there's an 4-year old issue for it: https://github.com/sorbet/sorbet/issues/2020
To be fair, it was fixed but then the fix was reverted.